### PR TITLE
Complete rewrite (Part II)

### PR DESCRIPTION
**Pull request by nwithan8**
This is a complete rewrite of Scotch (now called `EasyVCR`), introducing several new features and an overhauled (but incomplete) test suite.

Check the [updated README](https://github.com/EasyPost/scotch/blob/access_and_order/README.md) for details and [`Sample.cs`](https://github.com/EasyPost/scotch/blob/access_and_order/EasyPost.EasyVCR.Tests/Sample.cs) for examples.

This PR includes:
- Censoring sensitive information in bodies, queries and headers
- Order elements in cassette files (avoid git differences)
- Customize what defines a "match" when looking for a recorded request
- Simulate delays when replaying requests to better imitate real requests
- Auto mode (uses recording if exists, otherwise make a new recording)
- Set mode via environmental variable
- Locked down access to internal components to provide a black-box, "just works" experience
- Ability to override how elements are converted
- Sample examples in `Sample.cs` and README
- Auto-docs via DocFX

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,42 +1,138 @@
-# Scotch
+# EasyVCR
 
-### Disclaimer
-This is a fork of the original [Scotch by Martin Leech](https://github.com/mleech/scotch), updated to address outstanding functionality issues, as well as introduce a handful of new features. This version of Scotch is a C# rewrite of the original F# version.
+#### Disclaimer
```
Comment (Justintime50) [inline README.md:3]: Let's move this to the bottom of the README, more of an afterthought than the most exciting/important introduction to the project.

File: README.md [hunk 2]
```diff
@@ -1,42 +1,138 @@
...
+This is a distant fork of the original [Scotch by Martin Leech](https://github.com/mleech/scotch), completely rewritten in C# and enhanced with several new features and performance improvements.
 
-## What is Scotch?
-Scotch is a library for recording and replaying HTTP interactions in your test suite. 
+## What is EasyVCR?
```
Comment (Justintime50) [inline README.md:6]: Let's remove this title and have the next sentence/paragraph be the first thing after the header in the README

File: README.md [hunk 3]
```diff
@@ -1,42 +1,138 @@
...
+EasyVCR is a library for recording and replaying HTTP interactions in your test suite.
 
 This can be useful for speeding up your test suite, or for running your tests on a CI server which doesn't have connectivity to the HTTP endpoints you need to interact with.
 
-Scotch is based on the [VCR Ruby gem](https://github.com/vcr/vcr).
-
-### Step 1.
+## How to use EasyVCR
+#### Step 1.
```
Comment (Justintime50) [inline README.md:12]: Markdown likes a line break between stuff like this.

File: README.md [hunk 4]
```diff
@@ -1,42 +1,138 @@
...
+
+var cassette = new Cassette("path/to/cassettes", "my_cassette");
+var advancedOptions = new AdvancedOptions()
+{
+    Censors = new Censors().HideHeader("Authorization") // Hide the Authorization header
+};
+
+var httpClient = HttpClients.NewHttpClient(cassette, Mode.Record, advancedSettings);
+```
+### Delay
```
Comment (Justintime50) [inline README.md:70]: Line breaks before and after headings please

**Review by Justintime50 — COMMENTED** [review]
I think you need to gitignore the `docs` folder completely?

I'll take a look at this next week.

File: EasyPost.EasyVCR/Mode.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+{
+    /// <summary>
+    ///     Modes of operation for the VCR system.
+    /// </summary>
+    public enum Mode
+    {
+        /// <summary>
+        ///     No recording or playback. Make HTTP requests as normal.
+        /// </summary>
+        Bypass,
```
Comment (Justintime50) [inline EasyPost.EasyVCR/Mode.cs:11]: Just a note that `Bypass` doesn't really appear in the comment headers of other functions when the other 3 modes do. May be good to add it in those places?

File: EasyPost.EasyVCR/Mode.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+        /// <summary>
+        ///     Replay existing recordings if they exist, otherwise record new ones.
+        /// </summary>
+        Auto,
```
Comment (Justintime50) [inline EasyPost.EasyVCR/Mode.cs:15]: I can't tell yet, is `Auto` the default behavior? If not, it should be to match other VCR solutions.

Since this is a list of sorts, could we sort it alphabetically?

File: EasyPost.EasyVCR/Statics.cs [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        {
+            {
+                "Authorization"
+            }
+        };
+
+        /// <summary>
+        ///     Default list of parameters to censor in the cassettes.
+        /// </summary>
+        internal static List<string> DefaultCredentialParametersToHide => new()
```
Comment (Justintime50) [inline EasyPost.EasyVCR/Statics.cs:23]: When searching for creds to hide, do we lowercase them on matching? Otherwise this will fail with differing casings.

File: EasyPost.EasyVCR/Statics.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
+using System.Collections.Generic;
+
+namespace EasyPost.EasyVCR
+{
+    /// <summary>
+    ///     Static values used as defaults for EasyVCR.
+    /// </summary>
+    internal static class Statics
```
Comment (Justintime50) [inline EasyPost.EasyVCR/Statics.cs:8]: `Statics`... don't love the name. Are there better alternatives?

File: README.md [hunk 9]
```diff
@@ -1,52 +1,162 @@
...
+
+// Or make HTTP calls directly
+var response = await recordingHttpClient.GetAsync("https://api.example.com/v1/users");
 ```
 Real HTTP calls will be made and recorded to the cassette file.
 
-### Step 2.
+#### Step 2.
 Switch to replay mode:
 ```csharp
```
Comment (Justintime50) [inline README.md:35]: There should be line breaks between all these items. Double check the whole README for more of these too.

File: README.md [hunk 10]
```diff
@@ -1,52 +1,162 @@
...
+// create an HttpClient using the cassette
+var replayingHttpClient = HttpClients.NewHttpClient(cassette, Mode.Replay);
+```
+Now when tests are run, no real HTTP calls will be made. Instead, the HTTP responses will be replayed from the cassette file.
+
+### Available modes
+- `Mode.Record`: Record a request, including overwriting any existing matching recording.
+- `Mode.Replay`: Replay a request. Throws an exception if no matching recording is found.
+- `Mode.Auto`:  Play back a request if it has been recorded before, or record a new one if not.
+- `Mode.Bypass`:  Do not record or replay any requests (client will behave like a normal HttpClient).
```
Comment (Justintime50) [inline README.md:50]: Can we sort this alphabetically? Calling out the default (if there is one), if there isn't one we should set `Auto` has the default.

File: EasyPost.EasyVCR.Tests/ClientTest.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+            Assert.IsNotNull(client);
+        }
+
+        [TestMethod]
+        public async Task TestEraseAndPlayback()
+        {
+            var cassette = TestUtils.GetCassette("test_erase_and_playback");
+            cassette.Erase(); // Erase cassette before recording
+            Assert.IsTrue(cassette.Count == 0); // Make sure cassette is empty
```
[the remark this answers is not in the document]
Comment (nwithan8) [reply EasyPost.EasyVCR.Tests/ClientTest.cs:33]: I added `Count`, it tells you how many HTTP request/response pairs are present in any given cassette file.

File: EasyPost.EasyVCR.Tests/EasyPost.EasyVCR.Tests.csproj [hunk 12]
```diff
@@ -3,20 +3,24 @@
 <Project Sdk="Microsoft.NET.Sdk.Web">
     <PropertyGroup>
         <TargetFrameworks>net461;netstandard2.0;netcoreapp3.1;net5.0;net6.0</TargetFrameworks>
-        <LangVersion>latestMajor</LangVersion>
+        <LangVersion>9</LangVersion>
```
[the remark this answers is not in the document]
Comment (nwithan8) [reply EasyPost.EasyVCR.Tests/EasyPost.EasyVCR.Tests.csproj:6]: There's some stuff in here that only works on 9 and above. I refactored them to make them compatible with 8.

File: EasyPost.EasyVCR/AdvancedSettings.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        /// </summary>
+        public IInteractionConverter? InteractionConverter = new DefaultInteractionConverter();
+
+        /// <summary>
+        ///     Simulate the original request's duration. Overrides <see cref="ManualDelay"/> if set.
+        /// </summary>
+        public bool SimulateDelay = false;
+
+        /// <summary>
+        ///     Simulate a delay in milliseconds for the request.
```
[the remark this answers is not in the document]
Comment (nwithan8) [reply EasyPost.EasyVCR/AdvancedSettings.cs:34]: In case people want to make their HTTP playback as realistic as possible, testing timing, etc.
We probably won't use it for our immediate purposes (and it's off by default), but just a nice addition to have.

File: EasyPost.EasyVCR/Cassette.cs [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+        /// <summary>
+        ///     Erase this cassette by deleting the file
+        /// </summary>
+        public void Erase()
+        {
+            File.Delete(_filePath);
+        }
+
+        /// <summary>
+        ///     Lock this cassette (prevent reading or writing)
```
[the remark this answers is not in the document]
Comment (nwithan8) [reply EasyPost.EasyVCR/Cassette.cs:64]: So you don't accidentally record over something. Like the lock on a real-life cassette tape.

File: EasyPost.EasyVCR/CassetteOrder.cs [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+            /// <summary>
+            ///     Create an instance of the <see cref="Alphabetical"/> class.
+            /// </summary>
+            /// <param name="direction"></param>
+            public Alphabetical(Direction direction = Direction.Ascending)
+            {
+                _direction = direction;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (nwithan8) [reply EasyPost.EasyVCR/CassetteOrder.cs:68]: Must just be GitHub:
image

File: EasyPost.EasyVCR/Mode.cs (hunk 6 above)
Comment (nwithan8) [reply EasyPost.EasyVCR/Mode.cs:15]: Yes, `Auto` is the default for the VCR. For a standalone HttpClient, there is no default mode; you have to explicitly say what mode you want.

I'll sort the list.

File: EasyPost.EasyVCR/Statics.cs (hunk 7 above)
Comment (nwithan8) [reply EasyPost.EasyVCR/Statics.cs:23]: Good catch.

File: EasyPost.EasyVCR/Statics.cs (hunk 8 above)
Comment (Justintime50) [reply EasyPost.EasyVCR/Statics.cs:8]: I'm a fan of either, I'll leave the choice up to you!

File: README.md [hunk 16]
```diff
@@ -1,52 +1,162 @@
...
+var advancedOptions = new AdvancedOptions()
+{
+    InteractionConverter = new MyInteractionConverter(), // use a custom interaction converter by implementing IInteractionConverter
+};
+
+var httpClient = HttpClients.NewHttpClient(cassette, Mode.Replay, advancedSettings);
+```
+
+## VCR
+`EasyVCR` also offers a built-in VCR, which can be used to easily switch between multiple cassettes and/or modes. Any advanced settings applied to the VCR will be applied on every request made using the VCR's HttpClient.
```
[the remark this answers is not in the document]
Comment (nwithan8) [reply README.md:134]: I'll reword it to make it more clear. We can discuss it offline as well.

Basically, the original Scotch allowed you to make a cassette, and then pass that cassette into an `HttpClient`, and you could then use that `HttpClient` to make requests, and everything would be recorded to that cassette.

Scaling this us, say you want to use multiple cassettes. You'd have to constantly be remaking new `HttpClient` instances with the new cassette in them. Instead here, you can just change the cassette in the VCR, and grab the client from the VCR, which will give you what you need.

Effectively, the VCR maintains the current cassette and mode for you.

[the remark this answers is not in the document]
Comment (Justintime50) [reply README.md:134]: Ah thanks for clarifying. Yeah adding this helpful info to the project would be ideal.

**Review by Justintime50 — APPROVED** [review]

File: EasyPost.EasyVCR/RequestElements/HttpInteraction.cs [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+        ///     The HTTP request.
+        /// </summary>
+        [JsonProperty("Request")]
+        internal Request Request { get; set; }
+        /// <summary>
+        ///     The HTTP response.
+        /// </summary>
+        [JsonProperty("Response")]
+        internal Response Response { get; set; }
+
```
Comment (jchen293) [inline EasyPost.EasyVCR/RequestElements/HttpInteraction.cs:26]: remove this line here

**Review by jchen293 — APPROVED** [review]
lgtm
